File: backend/app/siem/schema/dashboards.py
from datetime import datetime
from typing import Any
from typing import Dict
from typing import List
from typing import Optional

from pydantic import BaseModel
from pydantic import Field

# ── Template browsing (read from disk) ───────────────────────────


class DashboardPanel(BaseModel):
    id: str
    title: str
    type: str
    w: int
    h: int
    lucene: str
    field: Optional[str] = None
    size: Optional[int] = None


class DashboardTemplate(BaseModel):
    id: str
    title: str
    description: str
    panels: List[DashboardPanel]


class DashboardCategory(BaseModel):
    id: str
    title: str
    description: str
    vendor: str
    product: str
    event_type: str
    tags: List[str]
    color: str
    icon: str


class DashboardCategoryWithTemplates(DashboardCategory):
    templates: List[DashboardTemplate]


class DashboardCategoriesListResponse(BaseModel):
    categories: List[DashboardCategory]
    success: bool
    message: str


class DashboardCategoryDetailResponse(BaseModel):
    category: DashboardCategoryWithTemplates
    success: bool
    message: str


# ── Enabled dashboards (DB-backed, per customer) ────────────────


class EnableDashboardRequest(BaseModel):
    customer_code: str = Field(..., max_length=50)
    event_source_id: int
    library_card: str = Field(..., max_length=255, description="Category id, e.g. 'wazuh_edr'")
    template_id: str = Field(..., max_length=255, description="Template id, e.g. 'EDR_OVERVIEW'")
    display_name: str = Field(..., max_length=255)


class EnabledDashboardResponse(BaseModel):
    id: int
    customer_code: str
    event_source_id: int
    library_card: str
    template_id: str
    display_name: str
    created_at: datetime

    class Config:
        orm_mode = True


class EnabledDashboardsListResponse(BaseModel):
    enabled_dashboards: List[EnabledDashboardResponse]
    success: bool
    message: str


class EnabledDashboardOperationResponse(BaseModel):
    enabled_dashboard: Optional[EnabledDashboardResponse] = None
    success: bool
    message: str


class DisableDashboardResponse(BaseModel):
    success: bool
    message: str


# ── Panel data (for rendering dashboards) ────────────────────────


class PanelDataRequest(BaseModel):
    dashboard_id: int
    timerange: str = Field("24h", description="Time range (e.g. '1h', '6h', '24h', '7d', '30d')")


class PanelResult(BaseModel):
    type: str
    value: Optional[int] = None  # stat panels
    labels: Optional[List[str]] = None  # chart panels
    data: Optional[List[Any]] = None  # chart panels
    error: Optional[str] = None


class PanelDataResponse(BaseModel):
    panels: Dict[str, PanelResult]
    template: DashboardTemplate
    dashboard_id: int
    customer_code: str
    source_name: str
    accent_color: str = "#38bdf8"
    success: bool
    message: str

File: frontend/src/types/dashboards.ts
export type DashboardPanelType = "stat" | "pie" | "bar_h" | "histogram" | "table"

export interface DashboardPanel {
	id: string
	title: string
	type: DashboardPanelType
	w: number
	h: number
	lucene: string
	/** Aggregation field, for `pie` / `bar_h` panels. */
	field?: string
	/** Source fields projected by `table` panels. */
	fields?: string[]
	size?: number
}

export interface DashboardTemplate {
	id: string
	title: string
	description: string
	panels: DashboardPanel[]
}

export interface DashboardCategory {
	id: string
	title: string
	description: string
	vendor: string
	product: string
	event_type: string
	tags: string[]
	color: string
	icon: string
}

export interface DashboardCategoryWithTemplates extends DashboardCategory {
	templates: DashboardTemplate[]
}

export interface EnabledDashboard {
	id: number
	customer_code: string
	event_source_id: number
	library_card: string
	template_id: string
	display_name: string
	created_at: string
}

export interface EnableDashboardPayload {
	customer_code: string
	event_source_id: number
	library_card: string
	template_id: string
	display_name: string
}

export interface PanelResult {
	type: string
	value: number | null
	labels: string[]
	data: number[]
	/** `table` panels only: column keys, in display order. */
	columns?: string[] | null
	/** `table` panels only: one object per document, keyed by column. */
	rows?: Record<string, string | number | boolean | null>[] | null
	error: string | null
}

export interface PanelDataResponse {
	panels: Record<string, PanelResult>
	template: DashboardTemplate
	dashboard_id: number
	customer_code: string
	source_name: string
	accent_color: string
}

// ── Custom dashboards (UI-authored, DB-backed templates) ─────────

/** Reserved `library_card` value used by every dashboard enabled from a custom template. */
export const CUSTOM_LIBRARY_CARD = "custom"

/** Panel as edited in the builder: the id is assigned by the backend when omitted. */
export interface CustomDashboardPanel extends Omit<DashboardPanel, "id"> {
	id?: string
}

/** Portable dashboard definition — the shape of an exported/imported JSON file. */
export interface CustomDashboardDefinition {
	template_key?: string | null
	title: string
	description: string
	vendor: string
	product: string
	event_type: string
	tags: string[]
	color: string
	icon: string
	default_query: string
	panels: CustomDashboardPanel[]
}

export interface CustomDashboard extends Omit<CustomDashboardDefinition, "template_key" | "panels"> {
	id: number
	template_key: string
	customer_code: string | null
	panels: DashboardPanel[]
	created_by: string | null
	created_at: string
	updated_at: string
}

export interface CustomDashboardCreatePayload extends CustomDashboardDefinition {
	customer_code?: string | null
}

export interface CustomDashboardUpdatePayload extends Partial<CustomDashboardDefinition> {
	customer_code?: string | null
	/** Explicit flag: a null `customer_code` can't express "share with everyone" in a partial update. */
	share_globally?: boolean
}

export interface CustomDashboardImportPayload {
	definition: CustomDashboardDefinition
	customer_code?: string | null
	overwrite?: boolean
}

export interface CustomDashboardPreviewPayload {
	event_source_id: number
	default_query: string
	panels: CustomDashboardPanel[]
	timerange: string
}

export interface CustomDashboardPreviewResponse {
	panels: Record<string, PanelResult>
	template: DashboardTemplate
	customer_code: string
	source_name: string
}
